airbyte-cdk/bulk/toolkits/load-gcs/src/main/kotlin/io/airbyte/cdk/load/file/gcs/GcsStreamingUpload.kt

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.gcs
+
+import com.google.cloud.storage.BlobId
+import com.google.cloud.storage.BlobInfo
+import com.google.cloud.storage.Storage
+import com.google.cloud.storage.Storage.ComposeRequest
+import io.airbyte.cdk.load.command.gcs.GcsClientConfiguration
+import io.airbyte.cdk.load.file.object_storage.StreamingUpload
+import io.airbyte.cdk.load.util.setOnce
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.concurrent.ConcurrentSkipListMap
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * This class is responsible for uploading individual parts to GCS and composing them at the end.
+ *
+ * /!\ Unlike typical "multipart upload" semantics, each part is visible as a standalone object
+ * until it is composed into a final object. That’s usually acceptable for BigQuery usage, but Will
+ * be an issue with a dedicated GCS destination as it would probably be a breaking change
+ */
+class GcsStreamingUpload(
+    private val storage: Storage,
+    private val key: String,
+    private val config: GcsClientConfiguration,
+    private val metadata: Map<String, String>
+) : StreamingUpload<GcsBlob> {
+
+    private val log = KotlinLogging.logger {}
+    private val isComplete = AtomicBoolean(false)
+    private val uploadId = generateUploadId()
+
+    /**
+     * Store part references: index => temporary object name. We do not store file contents in
+     * memory.
+     */
+    private val parts = ConcurrentSkipListMap<Int, String>()
+
+    /** Upload a part to GCS. Each part is stored as a temporary blob. */
+    override suspend fun uploadPart(part: ByteArray, index: Int) {
+        val partName = combinePath("$uploadId-part-$index")
+        log.info { "Uploading part #$index => $partName" }
+
+        val partBlobId = BlobId.of(config.gcsBucketName, partName)
+        val partBlobInfo = BlobInfo.newBuilder(partBlobId).build()
+
+        storage.create(partBlobInfo, part)
+
+        parts[index] = partName
+
+        log.info { "Uploaded part #$index => $partName, size: ${part.size} bytes" }
+    }
+
+    /** Compose all parts into the final object. */
+    override suspend fun complete(): GcsBlob {
+        if (!isComplete.setOnce()) {
+            log.warn { "Complete called multiple times for gs://${config.gcsBucketName}/$key" }
+            return GcsBlob(key, config)
+        }
+
+        if (parts.isEmpty()) {
+            // This is not ideal but follows what S3 is doing. Ideally we would return null
+            // or throw here. But until we update the CDK to handle that use case, this is
+            // better than creating an empty object in storage.
```

Comment:
not sure I follow? I feel like "create an empty blob" is a totally valid outcome. IMO "pretend we created a blob, even though we didn't" is more confusing, and certainly more surprising